File: src/core/vector.ts
/** Immutable 2D integer vector. Canvas cells are addressed by Vector. */
export class Vector {
  constructor(
    public readonly x: number,
    public readonly y: number,
  ) {}

  static fromKey(key: string): Vector {
    const i = key.indexOf(",");
    return new Vector(Number(key.slice(0, i)), Number(key.slice(i + 1)));
  }

  key(): string {
    return `${this.x},${this.y}`;
  }

  /** Display form, matching ASCIIFlow's `x:y`. */
  toString(): string {
    return `${this.x}:${this.y}`;
  }

  equals(other: Vector | null | undefined): boolean {
    return other != null && this.x === other.x && this.y === other.y;
  }

  add(other: Vector): Vector {
    return new Vector(this.x + other.x, this.y + other.y);
  }

  subtract(other: Vector): Vector {
    return new Vector(this.x - other.x, this.y - other.y);
  }

  scale(factor: number): Vector {
    return new Vector(this.x * factor, this.y * factor);
  }

  up(n = 1): Vector {
    return new Vector(this.x, this.y - n);
  }

  down(n = 1): Vector {
    return new Vector(this.x, this.y + n);
  }

  left(n = 1): Vector {
    return new Vector(this.x - n, this.y);
  }

  right(n = 1): Vector {
    return new Vector(this.x + n, this.y);
  }
}

/** The four unit directions. Instances are singletons, so compare with ===. */
export class Direction extends Vector {
  static readonly UP = new Direction(0, -1, "up");
  static readonly DOWN = new Direction(0, 1, "down");
  static readonly LEFT = new Direction(-1, 0, "left");
  static readonly RIGHT = new Direction(1, 0, "right");
  static readonly ALL: readonly Direction[] = [
    Direction.UP,
    Direction.DOWN,
    Direction.LEFT,
    Direction.RIGHT,
  ];

  private constructor(
    x: number,
    y: number,
    public readonly name: string,
  ) {
    super(x, y);
  }

  static of(x: number, y: number): Direction {
    const d = Direction.ALL.find((d) => d.x === x && d.y === y);
    if (!d) throw new Error(`Not a unit direction: ${x},${y}`);
    return d;
  }

  opposite(): Direction {
    switch (this) {
      case Direction.UP:
        return Direction.DOWN;
      case Direction.DOWN:
        return Direction.UP;
      case Direction.LEFT:
        return Direction.RIGHT;
      default:
        return Direction.LEFT;
    }
  }

  isHorizontal(): boolean {
    return this === Direction.LEFT || this === Direction.RIGHT;
  }
}

File: src/core/vector.rs
//! Grid geometry: cell positions and the four unit directions.

use std::fmt;

/// A canvas cell coordinate.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct Pos {
    pub x: i32,
    pub y: i32,
}

impl Pos {
    pub const fn new(x: i32, y: i32) -> Self {
        Self { x, y }
    }

    pub const fn offset(self, dx: i32, dy: i32) -> Self {
        Self {
            x: self.x + dx,
            y: self.y + dy,
        }
    }

    pub const fn add(self, other: Pos) -> Self {
        self.offset(other.x, other.y)
    }

    pub const fn subtract(self, other: Pos) -> Self {
        self.offset(-other.x, -other.y)
    }

    pub const fn scale(self, n: i32) -> Self {
        Self {
            x: self.x * n,
            y: self.y * n,
        }
    }

    pub const fn up(self) -> Self {
        self.offset(0, -1)
    }

    pub const fn down(self) -> Self {
        self.offset(0, 1)
    }

    pub const fn left(self) -> Self {
        self.offset(-1, 0)
    }

    pub const fn right(self) -> Self {
        self.offset(1, 0)
    }
}

/// Display form, matching ASCIIFlow's `x:y`.
impl fmt::Display for Pos {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}:{}", self.x, self.y)
    }
}

/// The four unit directions. Values, so they compare and hash by content.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum Direction {
    Up,
    Down,
    Left,
    Right,
}

impl Direction {
    pub const ALL: [Direction; 4] = [
        Direction::Up,
        Direction::Down,
        Direction::Left,
        Direction::Right,
    ];

    pub const fn delta(self) -> Pos {
        match self {
            Direction::Up => Pos::new(0, -1),
            Direction::Down => Pos::new(0, 1),
            Direction::Left => Pos::new(-1, 0),
            Direction::Right => Pos::new(1, 0),
        }
    }

    /// Connection bit for this direction's axis end.
    pub const fn bit(self) -> u8 {
        match self {
            Direction::Up => 1,
            Direction::Right => 2,
            Direction::Down => 4,
            Direction::Left => 8,
        }
    }

    pub const fn opposite(self) -> Direction {
        match self {
            Direction::Up => Direction::Down,
            Direction::Down => Direction::Up,
            Direction::Left => Direction::Right,
            Direction::Right => Direction::Left,
        }
    }

    pub const fn is_horizontal(self) -> bool {
        matches!(self, Direction::Left | Direction::Right)
    }

    /// `n` steps along this direction.
    pub const fn scale(self, n: i32) -> Pos {
        self.delta().scale(n)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn offsets_and_opposites() {
        let p = Pos::new(3, 4);
        assert_eq!(p.up().down(), p);
        assert_eq!(p.left().right(), p);
        assert_eq!(p.add(Direction::Right.scale(3)), Pos::new(6, 4));
        assert_eq!(p.to_string(), "3:4");
        assert_eq!(Direction::Up.opposite(), Direction::Down);
        assert_eq!(Direction::Up.bit(), 1);
        assert!(Direction::Left.is_horizontal());
    }
}
